src/stan/mcmc/hmc/nuts/base_nuts.hpp

```diff
@@ -70,83 +54,69 @@ namespace stan {
-        Eigen::VectorXd rho_minus(n_cont); rho_minus.setZero();
-
-        util.H0 = this->hamiltonian_.H(this->z_);
+        Eigen::VectorXd rho = this->z_.p;
+        double sum_weight = 1;
 
-        // Sample the slice variable
-        util.log_u = std::log(this->rand_uniform_());
-
-        // Build a balanced binary tree until the NUTS criterion fails
-        util.criterion = true;
-        int n_valid = 0;
+        double H0 = this->hamiltonian_.H(this->z_);
+        int n_leapfrog = 0;
+        double sum_metro_prob = 1;  // exp(H0 - H0)
 
+        // Build a trajectory until the NUTS criterion is no longer satisfied
         this->depth_ = 0;
         this->divergent_ = 0;
 
-        util.n_tree = 0;
-        util.sum_prob = 0;
+        while (this->depth_ < this->max_depth_) {
+          // Build a new subtree in a random direction
+          Eigen::VectorXd rho_subtree(rho.size());
+          rho_subtree.setZero();
 
-        while (util.criterion && (this->depth_ <= this->max_depth_)) {
-          // Randomly sample a direction in time
-          ps_point* z = 0;
-          Eigen::VectorXd* rho = 0;
+          bool valid_subtree = false;
+          double sum_weight_subtree = 0;
 
           if (this->rand_uniform_() > 0.5) {
-            z = &z_plus;
-            rho = &rho_plus;
-            util.sign = 1;
+            this->z_.ps_point::operator=(z_plus);
+            valid_subtree
+              = build_tree(this->depth_, rho_subtree, z_propose,
+                           H0, 1, n_leapfrog,
+                           sum_weight_subtree, sum_metro_prob, writer);
+            z_plus.ps_point::operator=(this->z_);
           } else {
-            z = &z_minus;
-            rho = &rho_minus;
-            util.sign = -1;
+            this->z_.ps_point::operator=(z_minus);
+            valid_subtree
+              = build_tree(this->depth_, rho_subtree, z_propose,
+                           H0, -1, n_leapfrog,
+                           sum_weight_subtree, sum_metro_prob, writer);
+            z_minus.ps_point::operator=(this->z_);
           }
 
-          // And build a new subtree in that direction
-          this->z_.ps_point::operator=(*z);
+          sum_weight += sum_weight_subtree;
+          if (!valid_subtree) break;
 
-          int n_valid_subtree = build_tree(depth_, *rho, 0, z_propose, util,
-                                           writer);
+          // Sample from an accepted subtree
           ++(this->depth_);
 
-          *z = this->z_;
-
-          // Metropolis-Hastings sample the fresh subtree
-          if (!util.criterion)
-            break;
-
-          double subtree_prob = 0;
-
-          if (n_valid) {
-            subtree_prob = static_cast<double>(n_valid_subtree) /
-              static_cast<double>(n_valid);
-          } else {
-            subtree_prob = n_valid_subtree ? 1 : 0;
-          }
-
-          if (this->rand_uniform_() < subtree_prob)
```

Comment:
If I'm following this, you're using an implicit binary search with a random prob in (0,1) at each branching point.  It should only require a single RNG that can then be used to guide the search down the subtrees.  I can't quite follow this closely enough to see if what you're doing then allows a subtree that isn't selected to be freed, which would be great.  I'd love to see the pseudo-code for this in the NUTS III paper.